compiler/front_end/generate_cached_parser.py

```diff
@@ -0,0 +1,256 @@
+
+    # Serialize the default errors map.
+    body.append(" defe = {\n")
+    for key, value in sorted(parser.default_errors.items()):
+        body += [f"  {key}:", sym(value), ",\n"]
+    body.append(" }\n")
+
+    # Finally, Parser construction.
+    body.append(
+        " return lr1.Parser(\n"  #
+        "  item_sets=None,\n"
+        "  goto=goto,\n"
+        "  action=act,\n"
+        f"  conflicts={repr(parser.conflicts)},\n"
+        "  terminals=None,\n"
+        "  nonterminals=None,\n"
+        "  productions=prods,\n"
+        "  default_errors=defe,\n"
+        " )"
+    )
+
+    # Postprocess the output to swap the symbol placeholders for proper
+    # symbols.
+    #
+    # Symbol assignment is done late (here) so that one-character symbols can
+    # be used for the most frequently-reference objects.
+    #
+    # It would be possible to inline objects that are only used once, but there
+    # do not seem to be very many of those, and it would be necessary to track
+    # symbol definitions individually, instead of just concatenating them all
+    # together in symbol_defs.
+    symbol_idents = {}  # Map of placeholder => final identifier
+    ident_counter = 0  # Counter used for generating identifiers
+    reserved_identifiers = "P S R A E prods act goto defe".split()
+
+    # Iterate through the symbols from most common to least common.
+    for _, _, symbol in sorted(
+        ((v, str(type(k)), k) for k, v in symbol_uses.items()), reverse=True
+    ):
+        # Find the next available identifier, skipping identifiers that are
+        # used by the skeleton and identifiers that are actually Python
+        # keywords.
+        while True:
+            ident = _identifier(ident_counter)
+            ident_counter += 1
+            if ident not in reserved_identifiers and not keyword.iskeyword(ident):
+                break
+        # Assign the final symbol to the placeholder.
+        symbol_idents[symbols[symbol]] = ident
+
+    # Iterate through the text fragments looking for any placeholders (S's),
+    # and replace the placeholders with their final symbol.
+    for l in (symbol_defs, body):
+        for i in range(len(l)):
+            if isinstance(l[i], S):
+                l[i] = symbol_idents[l[i]]
+
+    # Return the final result.
+    return "".join(header + symbol_defs + body)
+
+
+_HEADER = """
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from compiler.front_end import lr1
```

Comment:
Nit: maybe add a "this file is autogenerated" type disclaimer

Reply:
Done.